pkg/app/piped/appconfigreporter/appconfigreporter.go

```diff
@@ -0,0 +1,337 @@
+		}
+	}
+}
+
+// scanAppConfigs checks and reports two types of applications.
+// One is applications registered in Control-plane already, and another is ones that aren't registered yet.
+func (r *Reporter) scanAppConfigs(ctx context.Context) error {
+	if len(r.gitRepos) == 0 {
+		r.logger.Info("no repositories were configured for this piped")
+		return nil
+	}
+
+	// Make all repos up-to-date.
+	for repoID, repo := range r.gitRepos {
+		if err := repo.Pull(ctx, repo.GetClonedBranch()); err != nil {
+			r.logger.Error("failed to update repo to latest",
+				zap.String("repo-id", repoID),
+				zap.Error(err),
+			)
+			return err
+		}
+	}
+
+	// Create a map to determine from GitPath if the application is registered.
+	apps := r.applicationLister.List()
+	registeredAppPaths := make(map[string]struct{}, len(apps))
+	for _, app := range apps {
+		id := model.BuildGitPathID(app.GitPath.Repo.Id, app.GitPath.Path, app.GitPath.ConfigFilename)
+		registeredAppPaths[id] = struct{}{}
+	}
+
+	if err := r.updateUnregisteredApps(ctx, registeredAppPaths); err != nil {
+		return fmt.Errorf("failed to update unregistered applications: %w", err)
+	}
+	if err := r.updateRegisteredApps(ctx, registeredAppPaths); err != nil {
+		return fmt.Errorf("failed to update registered applications: %w", err)
+	}
+
+	return nil
+}
+
+// updateUnregisteredApps sends all unregistered application configurations to the control-plane.
+func (r *Reporter) updateUnregisteredApps(ctx context.Context, registeredAppPaths map[string]struct{}) error {
+	apps := make([]*model.ApplicationInfo, 0)
+	for repoID, repo := range r.gitRepos {
+		as, err := r.findUnregisteredApps(ctx, repo.GetPath(), repoID, registeredAppPaths)
+		if err != nil {
+			return err
+		}
+		apps = append(apps, as...)
+	}
+	if len(apps) == 0 {
+		return nil
+	}
+
+	_, err := r.apiClient.ReportUnregisteredApplicationConfigurations(
+		ctx,
+		&pipedservice.ReportUnregisteredApplicationConfigurationsRequest{
+			Applications: apps,
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to put the latest unregistered application configurations: %w", err)
+	}
+	return nil
+}
+
+// findUnregisteredApps finds out unregistered application info in the given git repository.
+func (r *Reporter) findUnregisteredApps(ctx context.Context, repoPath, repoID string, registeredAppPaths map[string]struct{}) ([]*model.ApplicationInfo, error) {
+	apps := make([]*model.ApplicationInfo, 0)
+	err := fs.WalkDir(r.fsys, strings.TrimPrefix(repoPath, "/"), func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return nil
+		}
+
+		if shouldSkip(repoID, repoPath, filepath.Base(path), registeredAppPaths, false) {
+			return nil
+		}
+
+		appInfo, err := r.readApplicationInfo(ctx, repoPath, path)
+		if err != nil {
+			r.logger.Error("failed to read application info",
+				zap.String("repo-id", repoID),
+				zap.String("config-file-path", path),
+				zap.Error(err),
+			)
+			return nil
```

Comment:
Returning nil because we want to return the found list as many as possible, right?
So let's add a comment to tell why we are returning nil instead of the received error.

Reply:
There you go.